src/neural/loader.cc

```diff
@@ -75,6 +139,10 @@ std::string DecompressGzip(const std::string& filename) {
   }
   gzclose(file);
 
+  if (filename == CommandLine::BinaryName()) {
+    return DecompressEmbedded(buffer);
```

Comment:
It just went through gunzip (which didn't do anything as it's known to be not a gz file), and now it's going to be gunzipped again?..
That's not nice.

Ideally we should have just one place where we decompress, whether we do that from the beginning or from some offset.

There is `gzdopen()` function which accepts file descriptor instead of filename.
I think for normal weights file we should open a file ourselves and pass it.
For reading the binary we should open, fseek() and also pass.

Not sure whether it will work on Windows though..

Reply:
Strangely enough, on windows it works fine, but on linux it decompresses fewer bytes than it should, without returnimg an error.

Reply:
This is the diff I'm trying, that works on windows but not on linux (with an embedded net, normal loading is fine):
```diff
diff --git a/src/neural/loader.cc b/src/neural/loader.cc
index acdfb01..adf54ee 100644
--- a/src/neural/loader.cc
+++ b/src/neural/loader.cc
@@ -113,8 +113,26 @@ std::string DecompressGzip(const std::string& filename) {
   int bytes_read = 0;
 
   // Read whole file into a buffer.
-  gzFile file = gzopen(filename.c_str(), "rb");
-  if (!file) throw Exception("Cannot read weights from " + filename);
+  FILE* fp = fopen(filename.c_str(), "rb");
+  if (!fp) {
+    throw Exception("Cannot read weights from " + filename);
+  }
+  if (filename == CommandLine::BinaryName()) {
+    int32_t size, magic;
+    if (fseek(fp, -8, SEEK_END) || fread(&size, 4, 1, fp) != 1 ||
+        fread(&magic, 4, 1, fp) != 1) {
+      throw Exception("Error reading embedded file descriptor.");
+    }
+    if (magic != 0x2130634c) {
+      throw Exception("No embedded file detected.");
+    }
+    fseek(fp, -size-8, SEEK_END);
+  }
+  gzFile file = gzdopen(fileno(fp), "rb");
+  if (!file) {
+    fclose(fp);
+    throw Exception("Cannot process file " + filename);
+  }
   while (true) {
     const int sz =
         gzread(file, &buffer[bytes_read], buffer.size() - bytes_read);
@@ -134,7 +152,7 @@ std::string DecompressGzip(const std::string& filename) {
   gzclose(file);
 
   if (filename == CommandLine::BinaryName()) {
-    return DecompressEmbedded(buffer);
+//    return DecompressEmbedded(buffer);
   }
 
   return buffer;
```
Any ideas?

Reply:
`fopen()` / `fclose()` are stdlib functions which have their own buffers. Probably it's not a good idea to steal a raw fileno from there.
I think `open()` / `close()` system calls are to be used instead. Not sure if they exist on windows, and whether libgzip does any buffering.

Reply:
The solution with `open()`/`close()` etc. works fine on linux but not on windows (while it compiles OK). However a `fflush()` does the trick, and `dup()` makes a new fileno.